src/AggregateFunctions/QuantileGK.h

```diff
@@ -0,0 +1,496 @@
+#pragma once
+
+#include <Common/RadixSort.h>
+#include <Common/PODArray.h>
+#include <IO/ReadBuffer.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteBuffer.h>
+#include <IO/WriteHelpers.h>
+#include <base/types.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int NOT_IMPLEMENTED;
+    extern const int INCORRECT_DATA;
+}
+
+
+template <typename T>
+class ApproxSampler
+{
+public:
+    struct Stats
+    {
+        T value; // the sampled value
+        Int64 g; // the minimum rank jump from the previous value's minimum rank
+        Int64 delta; // the maximum span of the rank
+
+        Stats() = default;
+        Stats(T value_, Int64 g_, Int64 delta_) : value(value_), g(g_), delta(delta_) { }
+    };
+
+    struct QueryResult
+    {
+        size_t index;
+        Int64 rank;
+        T value;
+
+        QueryResult(size_t index_, Int64 rank_, T value_) : index(index_), rank(rank_), value(value_) { }
+    };
+
+    ApproxSampler() = default;
+
+    ApproxSampler(const ApproxSampler & other)
+        : relative_error(other.relative_error)
+        , compress_threshold(other.compress_threshold)
+        , count(other.count)
+        , compressed(other.compressed)
+        , sampled(other.sampled.begin(), other.sampled.end())
+        , backup_sampled(other.backup_sampled.begin(), other.backup_sampled.end())
+        , head_sampled(other.head_sampled.begin(), other.head_sampled.end())
+    {
+    }
+
+    explicit ApproxSampler(
+        double relative_error_, size_t compress_threshold_ = default_compress_threshold, size_t count_ = 0, bool compressed_ = false)
+        : relative_error(relative_error_), compress_threshold(compress_threshold_), count(count_), compressed(compressed_)
+    {
+        sampled.reserve(compress_threshold);
+        backup_sampled.reserve(compress_threshold);
```

Comment:
Why do we need to reserve for `backup_sampled` here. Seems like a waste of memory. For small states reserving for `sampled` will be a waste of memory too (making -Merge use way more memory than necessary).

Reply:
It is used to avoid reallocs